src/core/qgspallabeling.cpp

```diff
@@ -1186,11 +1221,29 @@ void QgsPalLayerSettings::calculateLabelSize( const QFontMetricsF *fm, const QSt
 
 void QgsPalLayerSettings::registerFeature( const QgsFeature &f, QgsRenderContext &context, QgsLabelFeature **labelFeature, QgsGeometry obstacleGeometry )
 {
+  QgsFeature feature = f;
+  if ( geometryGeneratorEnabled )
+  {
+    if ( !mGeometryGeneratorExpression.isValid() )
+    {
+      mGeometryGeneratorExpression = QgsExpression( geometryGenerator );
+      mGeometryGeneratorExpression.prepare( &context.expressionContext() );
+      if ( mGeometryGeneratorExpression.hasParserError() )
+        QgsMessageLog::logMessage( QObject::tr( "Labeling" ), mGeometryGeneratorExpression.parserErrorString() );
+    }
+    context.expressionContext().setFeature( feature );
```

Comment:
This isn't required here -- the context will already have the incoming feature set (see qgsvectorlayerlabelprovider.cpp:250)